File: src/TCG/SentenceTranslator/FunretTranslator.cpp
#include "Common/Common.h"
#include "TCG/SentenceTranslator/FunretTranslator.h"
#include "TCG/ASM.h"
#include "TCG/SymbolManager.h"
#include "TCG/ConstructASM.h"

ASMLines FunretTranslator::SentenceTranslate_(SymbolManager& SymbolManager_, TACLine& TACLine_) {
    ASMLines asmlines;
    std::string ret_src = TACLine_.src1.value;
    int ret_mem = SymbolManager_.ret();
    if (TACLine_.src1.OperType == TACOPERANDTYPE::IMM) {
        asmlines.push_back(construct_asm("mov", ret_mem, ret_src));
    } else if (TACLine_.src1.OperType == TACOPERANDTYPE::VAR) {
        std::string ret_src_encode = SymbolManager_.encode_var(ret_src);
        POSTYPE pos = SymbolManager_.position(ret_src_encode);
        switch (pos) {
            case POSTYPE::REG: {
                REG ret_src_reg = SymbolManager_.avalue_reg(ret_src_encode);
                asmlines.push_back(construct_asm("mov", ret_mem, ret_src_reg));
                break;
            }
            case POSTYPE::MEM: {
                int ret_src_mem = SymbolManager_.avalue_mem(ret_src_encode);
                asmlines.push_back(construct_asm("mov", REG::EDI, ret_src_mem));
                asmlines.push_back(construct_asm("mov", ret_mem, REG::EDI));
                break;
            }
            case POSTYPE::GLOBAL: {
                asmlines.push_back(construct_asm("mov", ret_mem, ret_src));
                break;
            }
            default: {
                std::cout << "funret default error" << std::endl;
                break;
            }
        }
    }
    return asmlines;
}

File: go_compiler/tcg/sentence_translator/FunretTranslator.py
from loguru import logger
from ...common.REG import REG
from ...common.tac import TACOPERANDTYPE, TACLine
from ...tcg.construct_asm import construct_asm
from ...tcg.SymbolManager import POSTYPE, SymbolManager
from ...tcg.asm import ASMLines
from ...tcg.sentence_translator.BaseTranslator import BaseTranslator


class FunretTranslator(BaseTranslator):
    def __init__(self) -> None:
        super().__init__()

    def SentenceTranslate_(
        self, SymbolManager_: SymbolManager, TACLine_: TACLine
    ) -> ASMLines:
        asmlines: ASMLines = ASMLines()
        ret_src = TACLine_.src1.value
        ret_mem = SymbolManager_.ret()

        if TACLine_.src1.OperType == TACOPERANDTYPE.IMM:
            asmlines.append(construct_asm("mov", ret_mem, ret_src))
        elif TACLine_.src1.OperType == TACOPERANDTYPE.VAR:
            ret_src_encode = SymbolManager_.encode_var(ret_src)

            match SymbolManager_.position(ret_src_encode):
                case POSTYPE.REG:
                    ret_src_reg = SymbolManager_.avalue_reg(ret_src_encode)
                    asmlines.append(construct_asm("mov", ret_mem, ret_src_reg))
                case POSTYPE.MEM:
                    ret_src_mem = SymbolManager_.avalue_mem(ret_src_encode)
                    asmlines.append(construct_asm("mov", REG.EDI, ret_src_mem))
                    asmlines.append(construct_asm("mov", ret_mem, REG.EDI))
                case POSTYPE.GLOBAL:
                    asmlines.append(construct_asm("mov", ret_mem, ret_src))
                case _:
                    logger.error("funret default error")

        return asmlines
